language/move-lang/src/typing/core.rs

```diff
@@ -732,15 +732,42 @@ pub fn make_function_type(
     };
     let defined_loc = finfo.defined_loc;
     match &finfo.visibility {
-        FunctionVisibility::Internal if !in_current_module => {
-            let internal_msg = "This function is internal to its module. Only 'public' functions \
+        FunctionVisibility::Internal => {
+            if !in_current_module {
+                let internal_msg =
+                    "This function is internal to its module. Only 'public' functions \
                                 can be called outside of their module";
-            context.error(vec![
-                (loc, format!("Invalid call to '{}::{}'", m, f)),
-                (defined_loc, internal_msg.into()),
-            ])
+                context.error(vec![
+                    (loc, format!("Invalid call to '{}::{}'", m, f)),
+                    (defined_loc, internal_msg.into()),
+                ])
+            }
         }
-        _ => (),
+        FunctionVisibility::Script(_) => {
+            let in_script_context = match &context.current_module {
+                Some(current_m) => {
+                    let current_f = context
+                        .current_function
+                        .as_ref()
+                        .expect("ICE must be in a function to call other functions");
+                    let current_finfo = context.function_info(current_m, current_f);
+                    match &current_finfo.visibility {
+                        FunctionVisibility::Public(_) | FunctionVisibility::Internal => false,
+                        FunctionVisibility::Script(_) => true,
+                    }
+                }
+                None => true,
+            };
+            if !in_script_context {
+                let internal_msg = "This function can only be called from a script context, \
+                                i.e., a transaction script or a 'public(script)' function";
```

Comment:
nit, no comma after `i.e.` to be consistent with other stylings 

Also, we should really access `public(script)` via some constant